src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,686 @@
+_         * 0#        =  0#
+(2suc x)  * (2suc y)  =  2* (2suc (x + (y + x * y)))
+(2suc x)  * (suc2* y) =  2suc (x + y * (2suc x))
+                      --
+                      -- 2(1+x) * (1+2y) =  2(1 + 2y + x + 2xy)
+                      --                 =  2(1 + x + y*2(1 + x))
+
+(suc2* x) * (2suc y)  =  2suc (y + x * (2suc y))
+(suc2* x) * (suc2* y) =  suc2* (x + y * (suc2* x))
+            --
+            -- (1 + 2x)(1 + 2y) =  1 + (2y + 2x + 4xy)
+            --                     1 + 2(x + y * (1 + 2x))
+
+------------
+infixl 8 _^_
+
+_^_ : Bin → ℕ → Bin
+_ ^ 0      =  1#
+x ^ (1+ n) =  x * (x ^ n)
+
+------------------------------
+0* :  (x : Bin) → 0# * x ≡ 0#
+0* _ =  refl
+
+*0 :  (x : Bin) → x * 0#  ≡ 0#
+*0 0#        =  refl
+*0 (2suc _)  =  refl
+*0 (suc2* _) =  refl
+
+
+x*y≡0⇒⊎ :  {x y : Bin} → x * y ≡ 0# → x ≡ 0# ⊎ y ≡ 0#
+
+x*y≡0⇒⊎ {0#}      {_ }      _ =  inj₁ refl
+x*y≡0⇒⊎ {_}       {0#}      _ =  inj₂ refl
+x*y≡0⇒⊎ {2suc _}  {2suc _}  ()
+x*y≡0⇒⊎ {2suc _}  {suc2* _} ()
+x*y≡0⇒⊎ {suc2* _} {2suc _}  ()
+x*y≡0⇒⊎ {suc2* _} {suc2* _} ()
+
+nz*nz :  {x y : Bin} → x ≢ 0# → y ≢ 0# → x * y ≢ 0#
+nz*nz x≢0 y≢0 xy≡0  with x*y≡0⇒⊎ xy≡0
+...                 | inj₁ x≡0 =  x≢0 x≡0
+...                 | inj₂ y≡0 =  y≢0 y≡0
+
+
+setoid : Setoid 0ℓ 0ℓ
+setoid = PE.setoid Bin
+
+2*≢1 :  {x : Bin} → 2* x ≢ 1#
+2*≢1 {0#}      ()
+2*≢1 {2suc _}  ()
+2*≢1 {suc2* _} ()
+
+------------------------------------------------------------------------------
+suc≗1+ :  suc ≗ (1# +_)
+suc≗1+ 0#        =  refl
+suc≗1+ (2suc _)  =  refl
+suc≗1+ (suc2* _) =  refl
+
+------------------------------------------------------
+suc+assoc :  (x y : Bin) → (suc x) + y ≡ suc (x + y)
+
+suc+assoc 0# y =  begin suc 0# + y     ≡⟨⟩
+                        1# + y         ≡⟨ sym (suc≗1+ y) ⟩
+                        suc y          ≡⟨⟩
+                        suc (0# + y)
+                  ∎
+suc+assoc x 0# =  begin suc x + 0#     ≡⟨ +0 (suc x) ⟩
+                        suc x          ≡⟨ cong suc (sym (+0 x)) ⟩
+                        suc (x + 0#)
+                  ∎
+
+suc+assoc (2suc x) (2suc y) =
+  begin
+    (suc (2suc x)) + (2suc y)    ≡⟨⟩
+    suc2* (suc x) + (2suc y)     ≡⟨⟩
+    suc (2suc (suc x + y))       ≡⟨ cong (suc ∘ 2suc) (suc+assoc x y) ⟩
+    suc (2suc (suc (x + y)))     ≡⟨⟩
+    suc ((2suc x) + (2suc y))
+  ∎
+
+suc+assoc (2suc x) (suc2* y) =
+  begin
+    (suc (2suc x)) + (suc2* y)     ≡⟨⟩
+    (suc2* (suc x)) + (suc2* y)    ≡⟨⟩
+    suc (suc2* (suc x + y))        ≡⟨ cong (suc ∘ suc2*) (suc+assoc x y) ⟩
+    suc (suc2* (suc (x + y)))      ≡⟨⟩
+    suc (suc (2suc (x + y)))       ≡⟨⟩
+    suc ((2suc x) + (suc2* y))
+  ∎
+
+suc+assoc (suc2* x) (2suc y)  =  refl
+suc+assoc (suc2* x) (suc2* y) =  refl
+
+1+≗suc :  (1# +_) ≗ suc
+1+≗suc = suc+assoc 0#
+
+-------------
+infixr 7 2*'_
+
+toℕ : Bin → ℕ
+toℕ 0#        =  0
+toℕ (2suc x)  =  2 *' (1+ (toℕ x))
+toℕ (suc2* x) =  1+ (2 *' (toℕ x))
+
+
+2*'_ : ℕ → ℕ
+2*'_ = (2 *'_)
```

Comment:
I don't think saving one space character is worth it. Anyway this definitely shouldn't be publicly exported

Reply:
Improved to
```
private
   infixr 7 2*'_

   2*'_ : ℕ → ℕ
   2*'_ = (2 *'_)
```
This saves two parentheses in proofs.

Reply:
Sorry, I still don't think saving two parantheses is worth it.